File: scripts/beer_pack.py
"""Beer case/pack sizes from Supabase beer_pack_size table."""
from __future__ import annotations

import datetime as dt
from typing import Any

from clover_client import allowed_beer_line_items, get_config
from supabase_client import get_supabase

_CACHE: dict[str, tuple[float, dict[str, int]]] = {}
_CACHE_TTL_SEC = 300

# Fallback when DB row missing (matches original distributor rules)
_FALLBACK: dict[str, tuple[str, int]] = {
    "Miller Lite": ("bonbright", 12),
    "Guinness": ("bonbright", 12),
    "Blue Moon": ("bonbright", 12),
    "Coors Light": ("bonbright", 8),
    "Modelo": ("bonbright", 12),
    "Michelob Ultra": ("heidelberg", 24),
    "Yuengling": ("heidelberg", 24),
    "Bud Light": ("heidelberg", 24),
    "Angry Orchard": ("heidelberg", 24),
    "High Noon Pineapple": ("heidelberg", 24),
    "Busch Light": ("heidelberg", 24),
    "Truth": ("heidelberg", 24),
    "Boat Show (Yellow Springs)": ("yellow_springs", 12),
}


def _fallback_pack_size(beer_name: str) -> int:
    key = beer_name.strip()
    for name, (_, size) in _FALLBACK.items():
        if name.lower() == key.lower():
            return size
    return 12


def beer_pack_sizes_from_supabase(*, force_refresh: bool = False) -> dict[str, Any]:
    cfg = get_config()
    cache_key = cfg.merchant_id
    now = dt.datetime.now(dt.timezone.utc).timestamp()
    if not force_refresh and cache_key in _CACHE:
        ts, by_beer = _CACHE[cache_key]
        if now - ts < _CACHE_TTL_SEC:
            return {
                "ok": True,
                "merchant_id": cfg.merchant_id,
                "by_beer": by_beer,
                "from_cache": True,
            }

    by_beer: dict[str, int] = {}
    db_rows: dict[str, dict[str, Any]] = {}
    try:
        client = get_supabase()
        result = (
            client.table("beer_pack_size")
            .select("beer_name,distributor_id,pack_size")
            .eq("merchant_id", cfg.merchant_id)
            .execute()
        )
        for row in result.data or []:
            name = (row.get("beer_name") or "").strip()
            if name:
                db_rows[name.lower()] = row
    except Exception:
        db_rows = {}

    items_out: list[dict[str, Any]] = []
    for beer in allowed_beer_line_items():
        row = db_rows.get(beer.lower())
        if row:
            pack = int(row.get("pack_size") or 12)
            dist = row.get("distributor_id") or ""
            source = "database"
        else:
            dist, pack = _FALLBACK.get(beer, ("", 12))
            source = "fallback"
        by_beer[beer] = pack
        items_out.append(
            {
                "beer_name": beer,
                "distributor_id": dist,
                "pack_size": pack,
                "source": source,
            }
        )

    _CACHE[cache_key] = (now, by_beer)
    return {
        "ok": True,
        "merchant_id": cfg.merchant_id,
        "items": items_out,
        "by_beer": by_beer,
        "from_cache": False,
    }


def pack_size_for_beer(beer_name: str, *, force_refresh: bool = False) -> int:
    data = beer_pack_sizes_from_supabase(force_refresh=force_refresh)
    by_beer: dict[str, int] = data.get("by_beer") or {}
    for name, size in by_beer.items():
        if name.lower() == beer_name.strip().lower():
            return size
    return _fallback_pack_size(beer_name)

File: api/lib/beer-pack.js
/** Beer case sizes — Supabase beer_pack_size with fallback (sync with scripts/beer_pack.py). */
const FALLBACK = {
  "Miller Lite": 12,
  Guinness: 12,
  "Blue Moon": 12,
  "Coors Light": 8,
  Modelo: 12,
  "Michelob Ultra": 24,
  Yuengling: 24,
  "Bud Light": 24,
  "Angry Orchard": 24,
  "High Noon Pineapple": 24,
  "Busch Light": 24,
  Truth: 24,
  "Boat Show (Yellow Springs)": 12,
};

let cachedByBeer = null;
let cachedAt = 0;
const CACHE_MS = 5 * 60 * 1000;

function supabaseConfig() {
  const url = (process.env.SUPABASE_URL || "").trim().replace(/\/$/, "");
  const projectId = (process.env.SUPABASE_PROJECT_ID || "").trim();
  const base = url || (projectId ? `https://${projectId}.supabase.co` : "");
  const key =
    (process.env.SUPABASE_SERVICE_ROLE_KEY || "").trim() ||
    (process.env.SUPABASE_PUBLISHABLE_KEY || "").trim() ||
    (process.env.SUPABASE_ANON_KEY || "").trim();
  const merchantId = (process.env.CLOVER_MERCHANT_ID || "").trim();
  return { base, key, merchantId };
}

async function loadPackSizesFromSupabase() {
  const { base, key, merchantId } = supabaseConfig();
  if (!base || !key || !merchantId) return null;
  if (cachedByBeer && Date.now() - cachedAt < CACHE_MS) return cachedByBeer;

  const res = await fetch(
    `${base}/rest/v1/beer_pack_size?merchant_id=eq.${encodeURIComponent(merchantId)}&select=beer_name,pack_size`,
    {
      headers: {
        apikey: key,
        Authorization: `Bearer ${key}`,
        Accept: "application/json",
      },
    },
  );
  if (!res.ok) return null;
  const rows = await res.json();
  const byBeer = { ...FALLBACK };
  for (const row of rows) {
    const name = (row.beer_name || "").trim();
    if (name) byBeer[name] = Number(row.pack_size) || FALLBACK[name] || 12;
  }
  cachedByBeer = byBeer;
  cachedAt = Date.now();
  return byBeer;
}

export async function packSizeForBeerAsync(_distributorId, beerName) {
  const byBeer = (await loadPackSizesFromSupabase()) || FALLBACK;
  const key = Object.keys(byBeer).find((k) => k.toLowerCase() === (beerName || "").trim().toLowerCase());
  return key ? byBeer[key] : 12;
}

/** Sync fallback (matches Supabase seed). */
export function packSizeForBeer(distributorId, beerName) {
  const beer = (beerName || "").trim();
  const hit = Object.entries(FALLBACK).find(([k]) => k.toLowerCase() === beer.toLowerCase());
  if (hit) return hit[1];
  if (distributorId === "heidelberg") return 24;
  if (distributorId === "yellow_springs") return 12;
  return 12;
}

export { FALLBACK as BEER_PACK_FALLBACK };
